Makefile

```diff
@@ -0,0 +1,285 @@
+#   GOLDFLAGS: Extra linking flags passed to 'go' when building.
+#   GOGCFLAGS: Additional go compile flags passed to 'go' when building.
+#
+# Example:
+#   make
+#   make unittest
+#   make unittest GOFLAGS="-v -x"
+#   make unittest GOGCFLAGS="-N -l"
+#     Note: Use the -N -l options to disable compiler optimizations an inlining.
+#           Using these build options allows you to subsequently use source
+#           debugging tools like delve.
+endef
+.PHONY: unittest
+ifeq ($(HELP),y)
+unittest:
+	@echo "$$UNITTEST_HELP_INFO"
+else
+unittest: check-caimake
+	caimake go unittest
+endif 
+
+define BUILD_LOCAL_HELP_INFO
+# Build code on local.
+#
+# Args:
+#   WHAT: Directory names to build.  If any of these directories has a 'main'
+#     package, the build will produce executable files under bin/.
+#     If not specified, "everything" will be built.
+#   GOFLAGS: Extra flags to pass to 'go' when building.
+#   GOLDFLAGS: Extra linking flags passed to 'go' when building.
+#   GOGCFLAGS: Additional go compile flags passed to 'go' when building.
+#
+# Example:
+#   make
+#   make build-local
+#   make build-local WHAT=cmd/server GOFLAGS=-v
+#   make build-local GOGCFLAGS="-N -l"
+#     Note: Use the -N -l options to disable compiler optimizations an inlining.
+#           Using these build options allows you to subsequently use source
+#           debugging tools like delve.
+endef
+.PHONY: build-local
+ifeq ($(HELP),y)
+build-local:
+	@echo "$$BUILD_LOCAL_HELP_INFO"
+else
+build-local: check-caimake $(PRE_BUILD)
+	LOCAL_BUILD=true caimake go build $(WHAT) 
+endif 
+
+define BUILD_IN_CONTAINER_HELP_INFO
+# Build code in container.
+#
+# Args:
+#   WHAT: Directory names to build.  If any of these directories has a 'main'
+#     package, the build will produce executable files under bin/.
+#     If not specified, "everything" will be built.
+#   GOFLAGS: Extra flags to pass to 'go' when building.
+#   GOLDFLAGS: Extra linking flags passed to 'go' when building.
+#   GOGCFLAGS: Additional go compile flags passed to 'go' when building.
+#
+# Example:
+#   make
+#   make build-in-container
+#   make build-in-container WHAT=cmd/server GOFLAGS=-v
+#   make build-in-container GOGCFLAGS="-N -l"
+#     Note: Use the -N -l options to disable compiler optimizations an inlining.
+#           Using these build options allows you to subsequently use source
+#           debugging tools like delve.
+endef
+.PHONY: build-in-container
+ifeq ($(HELP),y)
+build-in-container:
+	@echo "$$BUILD_LINUX_HELP_INFO"
+else
+build-in-container: check-caimake $(PRE_BUILD)
+	LOCAL_BUILD=false caimake go build $(WHAT) 
+endif 
+
+define CONTAINER_HELP_INFO
+# Build docker image.
+#
+# Args:
+#   WHAT: Directories containing Dockerfile.
+#
+# Example:
+#   make container
+#   make container WAHT=build/server
+endef
+.PHONY: container 
+ifeq ($(HELP),y)
+container:
+	@echo "$$CONTAINER_HELP_INFO"
+else
+container: check-caimake
+	caimake docker build $(WHAT)
```

Comment:
So you call `caimake` here and in `caimake` binary, call bash script?  Why the turnaround?

Reply:
a binary is easier for distribution than a mess scripts.